corehq/apps/locations/queryutil.py

```diff
@@ -5,61 +5,105 @@
 from __future__ import unicode_literals
 
 import time
+import traceback
 from collections import defaultdict
 from contextlib import contextmanager
 
+from django.conf import settings
 from django.db.models.query import QuerySet
 
 from corehq.util.datadog.gauges import datadog_counter
 from corehq.util.datadog.utils import bucket_value
+from corehq.util.soft_assert import soft_assert
 
 
 class ComparedQuerySet(object):
-    """Measure times of MPTT queries
+    """Compare results and times of MPTT and CTE queries"""
 
-    Will be used to compare results and times of MPTT and CTE queries,
-    hence the name and slightly verbose implementation.
-    """
-
-    def __init__(self, mptt_set, timing_context):
+    def __init__(self, mptt_set, cte_set, timing_context):
         self._mptt_set = mptt_set
+        self._cte_set = cte_set
         if isinstance(timing_context, ComparedQuerySet):
             timing_context = timing_context._timing.clone()
         self._timing = timing_context
 
     def __str__(self):
-        return "MPTT query: {}".format(self._mptt_set.query)
+        return "MPTT query: {}\n\nCTE query: {}".format(
+            self._mptt_set.query,
+            self._cte_set.query,
+        )
 
     def __iter__(self):
+        items1 = []
+        finished = False
         with _commit_timing(self):
-            with self._timing("mptt") as timer1:
-                for item in self._mptt_set:
-                    with timer1.pause():
-                        yield item
+            try:
+                with self._timing("mptt") as timer1:
+                    for item in self._mptt_set:
+                        items1.append(item)
+                        with timer1.pause():
+                            yield item
+                finished = True
+            finally:
+                with self._timing("cte"):
+                    items2 = list(self._cte_set)
+                ids1 = {_identify(it) for it in items1}
+                ids2 = {_identify(it) for it in items2}
+                if (finished and ids1 != ids2) or not ids1.issubset(ids2):
+                    _report_diff(self, ids1, ids2, "" if finished else "incomplete iteration")
+                if finished:
+                    self.__len__()  # compares lengths -> reports diff if necessary
 
     def __len__(self):
         with _commit_timing(self):
             with self._timing("mptt"):
-                return len(self._mptt_set)
+                len1 = len(self._mptt_set)
+            with self._timing("cte"):
+                len2 = len(self._cte_set)
+        if len1 != len2:
+            ids1 = {_identify(it) for it in self._mptt_set}
+            ids2 = {_identify(it) for it in self._cte_set}
+            _report_diff(self, ids1, ids2, "%s != %s" % (len1, len2))
+        return len1
```

Comment:
if i'm understanding the circumstances under which _report_diff gets called, it's only checking that the length of each result is the same, and that the sets of ids are equal. that would mean that two different results like [1, 1, 2] and [1, 2, 2] would not get reported.

Reply:
That is correct in this case. I believe for the most part we expect there to be no duplicates like that. The reason I did the comparison this way is that the MPTT and CTE queries can have slightly different ordering. In most cases it doesn't matter, but the CTE ordering is better/more intuitive.